riverlog/river_log.go

```diff
@@ -0,0 +1,135 @@
+package riverlog
+
+import (
+	"bytes"
+	"cmp"
+	"context"
+	"encoding/json"
+	"errors"
+	"io"
+	"log/slog"
+	"os"
+
+	"github.com/riverqueue/river/internal/jobexecutor"
+	"github.com/riverqueue/river/rivertype"
+)
+
+const metadataKey = "river:log"
+
+const (
+	maxSizeMB    = 2
+	maxSizeBytes = maxSizeMB * 1024 * 1024
+)
+
+type contextKey struct{}
+
+// Logger extracts a logger from context from within the Work body of a worker.
+// Middleware must be installed on either the worker or client for this function
+// to be usable.
+func Logger(ctx context.Context) *slog.Logger {
+	logger, ok := ctx.Value(contextKey{}).(*slog.Logger)
+	if !ok {
+		panic("no logger in context; do you have riverlog.Middleware configured?")
+	}
+	return logger
+}
+
+type Middleware struct {
+	rivertype.Middleware
+	config     *MiddlewareConfig
+	newHandler func(w io.Writer) slog.Handler
+}
+
+type MiddlewareConfig struct {
+	// Logger is an internal logger for the middleware. This has no relation
+	// to the logger that's injected into work context and is only used to log
+	// warnings in cases like if log size exceeded the maximum allowed.
+	//
+	// Defaults to slog.TextHandler logging to stdout at warning level.
+	Logger *slog.Logger
+
+	// MaxSizeBytes is the maximum size of log data that'll be persisted in
+	// bytes per job attempt. Anything larger will be truncated will be
+	// truncated down to MaxSizeBytes.
+	//
+	// Be careful with this number because the maximum total log size is equal
+	// to maximum number of attempts multiplied by this number (each attempt's
+	// logs are kept separately). For example, 25 * 2 MB = 50 MB maximum
+	// theoretical log size. Log data goes into metadata which is a JSONB field,
+	// and JSONB fields have a maximum size of 255 MB, so any number larger than
+	// 255 divided by maximum number of attempts may cause serious operational
+	// problems.
+	//
+	// Defaults to 2 MB (which is per job attempt).
+	MaxSizeBytes int
```

Comment:
This made me wonder if somebody abusing this is likely to run into issues with the max number of bytes being sent in a single query (for the completer). I know there's a max param count but unsure about the bytes limit, we might want to look that up.

Reply:
I looked into this a bit and I think the limit is very large in practice. This answer indicates 1 GB:

https://dba.stackexchange.com/a/131425

I think a good improvement for this would be to give the metadata updater a way to append to array so that we could make sure that only one attempt's logs are being sent at a time. That'll definitely take some rework to make happen though.